app/src/main/java/com/sunasterisk/musixmatch/service/MusicService.java

```diff
@@ -0,0 +1,245 @@
+        if (mMediaPlayer != null) {
+            mOnMediaPlayerChangeListener.onMediaStateChange(true);
+            mMediaPlayer.start();
+        }
+    }
+
+    @Override
+    public void play(String uriString) {
+        release();
+        Uri uri = Uri.parse(uriString);
+        mMediaPlayer = MediaPlayer.create(mContext, uri);
+        mMediaPlayer.setOnCompletionListener(this);
+        mOnMediaPlayerChangeListener.onTrackChange(getCurrentTrack());
+        mOnMediaPlayerChangeListener.onMediaStateChange(true);
+        start();
+    }
+
+    @Override
+    public void previous() {
+        destroyMusic();
+        changePosition(PREVIOUS);
+        play(getCurrentUri());
+    }
+
+    @Override
+    public void next() {
+        destroyMusic();
+        changePosition(NEXT);
+        play(getCurrentUri());
+    }
+
+    @Override
+    public void onCompletion(MediaPlayer mp) {
+        onLoopModeChanged();
+    }
+
+    public MediaListener getMediaListener() {
+        return this;
+    }
+
+    private String getCurrentUri() {
+        return mTracks.get(mPosition).getData();
+    }
+
+    private void destroyMusic() {
+        if (mMediaPlayer != null) {
+            mMediaPlayer.stop();
+            mMediaPlayer.release();
+            mMediaPlayer = null;
+        }
+    }
+
+    /**
+     * <p> This method will check loop mode when current track has finished </p>
+     */
+    private void onLoopModeChanged() {
+        switch (mStateLoop) {
+            case LoopMode.LOOP_OFF:
+                if (mPosition < mTracks.size() - 1) {
+                    mPosition++;
+                    play(getCurrentUri());
+                } else {
+                    mPosition = 0;
+                    seek(0);
+                    stop();
+                    mOnMediaPlayerChangeListener.onMediaStateChange(false);
+                    mOnMediaPlayerChangeListener.onTrackChange(getCurrentTrack());
+                }
+                break;
+            case LoopMode.LOOP_ONE:
+                play(getCurrentUri());
+                break;
+            case LoopMode.LOOP_ALL:
+                if (mPosition == mTracks.size() - 1) {
+                    mPosition = 0;
+                } else {
+                    mPosition++;
+                }
+                play(getCurrentUri());
+                break;
+            case LoopMode.SHUFFLE_ON:
+                if (mPosition < mTracks.size()) {
+                    mPosition = mRandom.nextInt(mTracks.size() + 1);
+                }
+                play(getCurrentUri());
+                break;
+        }
+    }
+
+    @IntDef({PREVIOUS, NEXT})
+    @interface Action {
+    }
+
+    private void changePosition(@Action int actionValue) {
+        if (mStateLoop == LoopMode.SHUFFLE_ON) {
+            mPosition = mRandom.nextInt(mTracks.size() + 1);
+        } else if (mPosition == mTracks.size() - 1) {
+            mPosition = 0;
+        } else {
+            mPosition += actionValue;
```

Comment:
mPosition += actionValue; mà lớn hơn size của track thí sao?

Reply:
position nó không thể lớn hơn được size tracks đâu ạ.
ở trên em đã thực hiện Indef nó, nó chỉ nhân -1 or +1,  và else if (mPosition == mTracks.size() - 1) {
            mPosition = 0;
nên là nó lên đến =track.size-1 thì nó đã về 0 rồi ạ

Reply:
nếu muốn an toàn thì e có thể cho sửa esle if ở trên thành >=

Reply:
đoạn code này e code rối quá em ạ
bh e chỉ cần viết 2 method
increatePosition và decreasePosition là xong.
ntn khó đọc khso maintain